### PR TITLE
Add grayscaleToRGB ops

### LINKED ISSUE #5134: tf.image.greyscale_to_rgb
Useful for converting greyscale image back to rgb so that the output can be rendered on canvas with toPixels.

tf api: https://www.tensorflow.org/api_docs/python/tf/image/grayscale_to_rgb

### PR DESCRIPTION
**Pull request by raffizulvian**
Adapt [tf.image.grayscale_to_rgb](https://www.tensorflow.org/api_docs/python/tf/image/grayscale_to_rgb) API to tfjs. Already create kernel for CPU and WebGL backend.

Close #5134

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)

### CODE REVIEW

**Comment by raffizulvian** [comment]
Hi @lina128 , can you please take a look? Thank you

**Review by lina128 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @raffizulvian)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 30 at r2](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MekR7PbBYqQ7IontIB2:-MekR7PbBYqQ7IontIB3:bu4l6j0) ([raw file](https://github.com/tensorflow/tfjs/blob/b24284875b1a3fbe67b20afbe2550c6bf8a05575/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L30)):*
> ```
>  * Converts images from grayscale to RGB by expanding the input image last
>  * dimension size from 1 into 3.
> ```
It can just be `Converts images from grayscale to RGB.` The rest of the sentence is implementation details, so should be hidden from doccomment. Also general suggestion for the first paragraph of doccomment is to complete the sentence in one line.
___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 55 at r2](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MekRduM0k34Cus2C4kB:-MekRduM0k34Cus2C4kC:bxnw742) ([raw file](https://github.com/tensorflow/tfjs/blob/b24284875b1a3fbe67b20afbe2550c6bf8a05575/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L55)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
>   const inputs: GrayscaleToRGBInputs = {image: $image};
>   const res = ENGINE.runKernel(
>     GrayscaleToRGB, inputs as {} as NamedTensorMap, {}
>   );
>   return res as T;
> ```
</details>

I think this op doesn't need to be implemented at kernel level, because it basically is a composition of several other ops, so it can just be implemented at op level, follow the implementation here: https://github.com/tensorflow/tensorflow/blob/8106d4b5fa426a09c205b38d2ded8a6ac57834be/tensorflow/python/ops/image_ops_impl.py#L2519-L2550

So basically you tile the last dimension three times: `const res = tile($image, [1, 1, 1, 3]);`

**Comment by raffizulvian** [comment]
> It can just be `Converts images from grayscale to RGB.` The rest of the sentence is implementation details, so should be hidden from doccomment. Also general suggestion for the first paragraph of doccomment is to complete the sentence in one line.

Oh okay, I will fix it.

> I think this op doesn't need to be implemented at kernel level, because it basically is a composition of several other ops, so it can just be implemented at op level, follow the implementation here

Working on it.

**Comment by raffizulvian** [comment]
I have a question, should this ops just cover only rank 4 tensor or tensor with rank more than and equal to 2 (rank >= 2) too? @lina128

**Review by raffizulvian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @lina128)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 30 at r2](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MekR7PbBYqQ7IontIB2:-MeqE2De2S3__fhtOGZP:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/8ede966f58d55699d61937e191a095cda4686679/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L30)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
>  * Converts images from grayscale to RGB by expanding the input image last
>  * dimension size from 1 into 3.
> ```
It can just be `Converts images from grayscale to RGB.` The rest of the sentence is implementation details, so should be hidden from doccomment. Also general suggestion for the first paragraph of doccomment is to complete the sentence in one line.
</blockquote></details>

Done.
___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 55 at r2](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MekRduM0k34Cus2C4kB:-MeqDwnE4QCZ3W_-2scR:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/8ede966f58d55699d61937e191a095cda4686679/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L55)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
>   const inputs: GrayscaleToRGBInputs = {image: $image};
>   const res = ENGINE.runKernel(
>     GrayscaleToRGB, inputs as {} as NamedTensorMap, {}
>   );
>   return res as T;
> ```
I think this op doesn't need to be implemented at kernel level, because it basically is a composition of several other ops, so it can just be implemented at op level, follow the implementation here: https://github.com/tensorflow/tensorflow/blob/8106d4b5fa426a09c205b38d2ded8a6ac57834be/tensorflow/python/ops/image_ops_impl.py#L2519-L2550

So basically you tile the last dimension three times: `const res = tile($image, [1, 1, 1, 3]);`
</blockquote></details>

Done.

**Comment by raffizulvian** [comment]
Sorry, I accidentally deleted my comment yesterday. Should this op be implemented for only rank 4 tensors? Or all tensors with at least rank 2 like in python API?

**Review by lina128 — COMMENTED** [review]
It should behave the same as python API, so rank 2, 3, and 4?

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308#-:-MezoRcy-xECXTHJauTh:b-5v2fue)* status: 0 of 1 approvals obtained (waiting on @lina128 and @raffizulvian)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 50 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MezoghH9qrI4EhgyPUm:-MezoghH9qrI4EhgyPUn:bblajsg) ([raw file](https://github.com/tensorflow/tfjs/blob/c6240107334a4e265e8273a5c85e3984209ef772/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L50)):*
> ```
> const result = tile($image, [1, 1, 1, 3])
> ```
You can just return at this line: `return tile(...`

**Review by raffizulvian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @lina128)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 50 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MezoghH9qrI4EhgyPUm:-Mf3_R0G-S0_hBVo6xUW:b16o0vh) ([raw file](https://github.com/tensorflow/tfjs/blob/c6240107334a4e265e8273a5c85e3984209ef772/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L50)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> const result = tile($image, [1, 1, 1, 3])
> ```
You can just return at this line: `return tile(...`
</blockquote></details>

After rechecking, the [python API implementation](https://github.com/tensorflow/tensorflow/blob/8106d4b5fa426a09c205b38d2ded8a6ac57834be/tensorflow/python/ops/image_ops_impl.py#L275-L277) of this op could accept rank 5, 6, and even 8. So, I'm not limiting the input to be only tensors with rank 2, 3, or 4. What do you think about this?

The only constraints are the input at least should have a two-dimensional shape (rank 2) and its shape last dimension should be 1.

**Review by lina128 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @lina128 and @raffizulvian)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 50 at r3](https://reviewable.io/reviews/tensorflow/tfjs/5308#-MezoghH9qrI4EhgyPUm:-Mf3j3UE9OjNWLRQQaIb:b32w49) ([raw file](https://github.com/tensorflow/tfjs/blob/c6240107334a4e265e8273a5c85e3984209ef772/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L50)):*
<details><summary><i>Previously, raffizulvian (Raffi Zulvian Muzhaffar) wrote…</i></summary><blockquote>

After rechecking, the [python API implementation](https://github.com/tensorflow/tensorflow/blob/8106d4b5fa426a09c205b38d2ded8a6ac57834be/tensorflow/python/ops/image_ops_impl.py#L275-L277) of this op could accept rank 5, 6, and even 8. So, I'm not limiting the input to be only tensors with rank 2, 3, or 4. What do you think about this?

The only constraints are the input at least should have a two-dimensional shape (rank 2) and its shape last dimension should be 1.
</blockquote></details>

SG
___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 54 at r4](https://reviewable.io/reviews/tensorflow/tfjs/5308#-Mf3k5UK0ICfPXY3EPod:-Mf3k5UL6K0Cr4mr9Sbl:b-jb5p9o) ([raw file](https://github.com/tensorflow/tfjs/blob/c1088571e50800a677bece8269cef3b211904df7/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L54)):*
> ```
>   for (let i = 0; i < lastDimsIdx; i++) {
>     reps[i] = 1;
>   }
> ```
Can this be: `reps.fill(1, 0, lastDimsIdx)`
___
*[tfjs-core/src/ops/image/grayscale_to_rgb_test.ts, line 65 at r4](https://reviewable.io/reviews/tensorflow/tfjs/5308#-Mf3jMNCBd3SZfge5SVw:-Mf3jMNCBd3SZfge5SVx:bsz2m6o) ([raw file](https://github.com/tensorflow/tfjs/blob/c1088571e50800a677bece8269cef3b211904df7/tfjs-core/src/ops/image/grayscale_to_rgb_test.ts#L65)):*
> ```
> expectArraysClose(rgbData, expected);
> ```
Test shape too.

**Review by raffizulvian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @lina128 and @raffizulvian)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 54 at r4](https://reviewable.io/reviews/tensorflow/tfjs/5308#-Mf3k5UK0ICfPXY3EPod:-Mf3zFma578SGNybAH9H:b-s3xqvb) ([raw file](https://github.com/tensorflow/tfjs/blob/c1088571e50800a677bece8269cef3b211904df7/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L54)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
>   for (let i = 0; i < lastDimsIdx; i++) {
>     reps[i] = 1;
>   }
> ```
Can this be: `reps.fill(1, 0, lastDimsIdx)`
</blockquote></details>

Working, will do
___
*[tfjs-core/src/ops/image/grayscale_to_rgb_test.ts, line 65 at r4](https://reviewable.io/reviews/tensorflow/tfjs/5308#-Mf3jMNCBd3SZfge5SVw:-Mf3zPyw1YVYKJTA_JZC:b-s3xqvb) ([raw file](https://github.com/tensorflow/tfjs/blob/c1088571e50800a677bece8269cef3b211904df7/tfjs-core/src/ops/image/grayscale_to_rgb_test.ts#L65)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> expectArraysClose(rgbData, expected);
> ```
Test shape too.
</blockquote></details>

Working, will do

**Review by lina128 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @lina128 and @raffizulvian)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 51 at r5](https://reviewable.io/reviews/tensorflow/tfjs/5308#-Mf4PBSzCF1RzpUiGyCt:-Mf4PBT-7MoCfoa5BQ4e:b-aqfm1n) ([raw file](https://github.com/tensorflow/tfjs/blob/1fbd5a113fa8210671c45e1651906d838ef01790/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L51)):*
> ```
> Uint8Array
> ```
Why use Uint8Array instead of Array?

**Review by raffizulvian — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308)* status: 0 of 1 approvals obtained (waiting on @lina128 and @raffizulvian)

___
*[tfjs-core/src/ops/image/grayscale_to_rgb.ts, line 51 at r5](https://reviewable.io/reviews/tensorflow/tfjs/5308#-Mf4PBSzCF1RzpUiGyCt:-Mf4QRJf8EIjjxRhUm5G:b-2g9qu6) ([raw file](https://github.com/tensorflow/tfjs/blob/1fbd5a113fa8210671c45e1651906d838ef01790/tfjs-core/src/ops/image/grayscale_to_rgb.ts#L51)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

> ```
> Uint8Array
> ```
Why use Uint8Array instead of Array?
</blockquote></details>

Oh, my ts gives an error earlier. I think it is because I'm not updating the branch. No more errors now after update the branch.

**Review by lina128 — APPROVED** [review]
Thank you for the contribution, @raffizulvian, great work!

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5308#-:-Mf4UKNe48oDNBjOSTjI:b-caslk)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @lina128 and @raffizulvian)

**Comment by raffizulvian** [comment]
Thank you so much for the guidance too! @lina128
